plugin/pkg/scheduler/algorithm/scheduler_interface.go

```diff
@@ -18,10 +18,24 @@ package algorithm
 
 import (
 	"k8s.io/kubernetes/pkg/api"
+	schedulerapi "k8s.io/kubernetes/plugin/pkg/scheduler/api"
 )
 
+// SchedulerExtender is an abstract interface for external processes to influence scheduling
+// decisions made by Kubernetes.
+type SchedulerExtender interface {
+	// Filter based on extender-implemented predicate functions. The filtered list is
+	// expected to be a subset of the supplied list.
+	Filter(pod *api.Pod, nodes *api.NodeList) (*api.NodeList, error)
+
+	// Prioritize based on extender-implemented priority functions. The returned scores & weight
+	// are used to compute the weighted score for an extender. The weighted scores are added to
+	// the scores computed  by Kubernetes scheduler. The total scores are used to do the host selection.
+	Prioritize(pod *api.Pod, nodes *api.NodeList) (*schedulerapi.HostPriorityList, int, error)
+}
+
 // Scheduler is an interface implemented by things that know how to schedule pods
 // onto machines.
 type ScheduleAlgorithm interface {
-	Schedule(*api.Pod, NodeLister) (selectedMachine string, err error)
+	Schedule(*api.Pod, NodeLister, []SchedulerExtender) (selectedMachine string, err error)
```

Comment:
> This is assuming all schedule algorithms takes extensions

No; it can be empty.

> not all schedule algorithms has the same extension interface like filter and prioritize

I think this extension mechanism is just intended to work with schedulers that use scheduler.genericScheduler, hence they do use predicate and priority functions. I think this extension approach is reasonable given how things work today.